Bigtable/src/Connection/Grpc.php

```diff
@@ -429,87 +407,164 @@ public function generateConsistencyToken(array $args)
      */
     public function snapshotTable(array $args)
     {
-        throw new \BadMethodCallException('This method is not implemented yet');
+        if (isset($args['ttl'])) {
+            $args['ttl'] = $this->serializer->decodeMessage(
+                new Duration,
+                $args['ttl']
+            );
+        }
+
+        $name = $this->pluck('name', $args);
+        $response = $this->send([$this->bigtableTableAdminClient, 'snapshotTable'], [
+            $name,
+            $this->pluck('cluster', $args),
+            $this->pluck('snapshotId', $args),
+            $this->pluck('description', $args),
+            $this->addResourcePrefixHeader($args, $name)
+        ]);
+
+        return $this->operationToArray(
+            $response,
+            $this->serializer,
+            $this->lroResponseMappers
+        );
     }
 
     /**
      * @param array $args
      */
     public function getSnapshot(array $args)
     {
-        throw new \BadMethodCallException('This method is not implemented yet');
+        $name = $this->pluck('name', $args);
+        return $this->send([$this->bigtableTableAdminClient, 'getSnapshot'], [
+            $name,
+            $this->addResourcePrefixHeader($args, $name)
+        ]);
     }
 
     /**
      * @param array $args
      */
     public function listSnapshots(array $args)
     {
-        throw new \BadMethodCallException('This method is not implemented yet');
+        $parent = $this->pluck('parent', $args);
+        return $this->send([$this->bigtableTableAdminClient, 'listSnapshots'], [
+            $parent,
+            $this->addResourcePrefixHeader($args, $parent)
+        ]);
     }
 
     /**
      * @param array $args
      */
     public function deleteSnapshot(array $args)
     {
-        throw new \BadMethodCallException('This method is not implemented yet');
-    }
-
-    /**
-     * @param array $args
-     */
-    public function readRow(array $args)
-    {
-        throw new \BadMethodCallException('This method is not implemented yet');
+        $name = $this->pluck('name', $args);
+        return $this->send([$this->bigtableTableAdminClient, 'deleteSnapshot'], [
+            $name,
+            $this->addResourcePrefixHeader($args, $name)
+        ]);
     }
 
     /**
      * @param array $args
      */
     public function readRows(array $args)
     {
-        throw new \BadMethodCallException('This method is not implemented yet');
+        if (isset($args['rows'])) {
+            $args['rows'] = $this->rowSetObject($args['rows']);
+        }
+
+        if (isset($args['filter'])) {
+            $args['filter'] = $this->rowFilterObject($args['filter']);
+        }
+
+        $name = $this->pluck('tableName', $args);
+        return $this->send([$this->bigtableClient, 'readRows'], [
```

Comment:
This works for now, but there's should probably be a TODO.  We need a couple of things for this to be useful:

1) a translation from a `ReadRowsResponse` to some form of a `Row` representation.
2) "smart" retries that allow users to retry from after the `Row` previously sent to a user.

Reply:
I believe we had some similar requirements in Spanner. Where we ended up putting this logic was in the "handwritten" layer ([Google\Cloud\Spanner\Result](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/Spanner/src/Result.php)). Would you see an advantage to this being in the connection?